pyani_plus/methods/method_anim.py

```diff
@@ -0,0 +1,254 @@
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+#
+"""Code to implement the ANIm average nucleotide identity method.
+
+Calculates ANI by the ANIm method, as described in Richter et al (2009)
+Proc Natl Acad Sci USA 106: 19126-19131 doi:10.1073/pnas.0906412106.
+
+All input FASTA format files are compared against each other, pairwise,
+using NUCmer (binary location must be provided). NUCmer output will be stored
+in a specified output directory.
+
+The NUCmer .delta file output is parsed to obtain an alignment length
+and similarity error count for every unique region alignment. These are
+processed to give matrices of aligned sequence lengths, similarity error
+counts, average nucleotide identity (ANI) percentages, and minimum aligned
+percentage (of whole genome) for each pairwise comparison.
+"""
+
+from collections import defaultdict
+from pathlib import Path
+from typing import NamedTuple
+
+import intervaltree  # type: ignore  # noqa: PGH003
+from Bio import SeqIO  # type: ignore  # noqa: PGH003
+
+from pyani_plus import utils
+from pyani_plus.tools import get_nucmer
+
+
+class ComparisonResult(NamedTuple):
+    """A single nucmer comparison result for ANIm.
+
+    We use a NamedTuple rather than a Dataclass because the
+    anim result should be immutable (we don't need to change it),
+    and we don't need to add any additional methods, or set
+    default values.
+    """
+
+    qhash: str  # query MD5 hash name
+    rhash: str  # reference MD5 hash name
+    q_aligned_bases: int  # aligned base count of query sequence
+    r_aligned_bases: int  # aligned base count of reference sequence
+    sim_errs: int  # count of similarity errors
+    avg_id: float  # average nucleotide identity (as a percentage)
+    program: str  # the program used to calculate the comparison
+    q_length: int  # total base count of query sequence
+    r_length: int  # total base count of reference sequence
+    q_cov: float  # query coverage (as a percentage)
+    r_cov: float  # reference coverage (as a percentage)
+    r_hadamard: float  # reference hadamard (percentage identity * percentage coverage)
+    q_hadamard: float  # query hadamard (percentage identity * percentage coverage)
+
+
+# NOTE This might not be needed as this information can now
+# be stored in one of the dataframeas in the database.
+def get_genome_length(filename: Path) -> int:
```

Comment:
You shouldn't need this but it will need some rearranging to have access to the DB at the time the value is used.

Reply:
Do you think that should be addressed under this branch now, or perhaps under a new one?

Reply:
Probably a new PR - I suspect it will need some refactoring.